### ISSUE TITLE
Alter table hangs on add/drop column

### ISSUE THREAD

**Issue by mohsenasm** [cockroachdb/cockroach#111203]
**Describe the problem**

An alter table hung for hours (on a small table) until I stopped the shell.

```sql
ALTER TABLE db_name.public.table_name ADD COLUMN column_name_2 UUID NOT NULL DEFAULT gen_random_uuid()
```

I tried `show jobs` and  `cancel job 902973788233302017`.
The result of `show jobs`:

```
        job_id       |   job_type    |                                                 description                                            | statement | user_name |  status   | running_status |             created              |             started              | finished |             modified             | fraction_completed |        error         | coordinator_id
---------------------+---------------+--------------------------------------------------------------------------------------------------------+-----------+-----------+-----------+----------------+----------------------------------+----------------------------------+----------+----------------------------------+--------------------+----------------------+-----------------
  902973788233302017 | SCHEMA CHANGE | ALTER TABLE db_name.public.table_name ADD COLUMN column_name_2 UUID NOT NULL DEFAULT gen_random_uuid() |           | root      | reverting | NULL           | 2023-09-25 10:02:11.117015+00:00 | 2023-09-25 10:02:11.815853+00:00 | NULL     | 2023-09-25 10:02:11.117015+00:00 |                  0 | job canceled by user |           NULL
  869008109992148993 | SCHEMA CHANGE | ALTER TABLE db_name.public.table_name DROP COLUMN column_name_1                                        |           | root      | reverting | NULL           | 2023-05-28 10:43:48.844502+00:00 | 2023-05-28 10:43:49.743624+00:00 | NULL     | 2023-05-28 10:43:48.844502+00:00 |                  0 | job canceled by user |              0
(2 rows)
```

This issue also happened 4 months ago with the same table (but for a different column). Between these two issues, there was no other issue with schema change in this table or other tables.

After canceling the job, the alter table returns this error:
```
root@:26257/db_name> ALTER TABLE table_name ADD COLUMN column_name_2 UUID NOT NULL DEFAULT gen_random_uuid();
ERROR: duplicate: column "column_name_2" in the middle of being added, not yet public
SQLSTATE: 42701
```

I don't know how I can get rid of the `reverting` jobs. Restarting the cluster didn't help.

**Expected behavior**

Change the table's schema or `revert` the job.

**Environment:**
 - CockroachDB version 20.2.4
 - Server OS: Docker cockroachdb/cockroach:v20.2.4
 - Client app: cockroach sql

Potentially related issues:
 - https://github.com/cockroachdb/cockroach/issues/13380
 - https://github.com/cockroachdb/cockroach/issues/12123
 - https://github.com/cockroachdb/cockroach/pull/70334

Jira issue: CRDB-31819

**Comment by msbutler** [comment]
:wave @mohsenasm Unfortunately we no longer support crdb version 20.2. Is there any way you could update your deployment to a new version, which has addressed many job contention problems.

**Comment by mohsenasm** [comment]
TL;DR
I had to drop & recreate the table to fix the issue.

------------------

More details for the next viewers:

I updated from `v20.2.4` to `v21.1.21` to `v21.2.17` to `v22.1.22` to `v22.2.14` to `v23.1.11`.
In the `v21.2.17` to `v22.1.22` step, the problem appeared again and prevented the update finalization:

```
root@:26257/defaultdb> SET CLUSTER SETTING version = "22.1";
ERROR: verifying precondition for version 21.2-2: there exists invalid descriptors as listed below; fix these descriptors before attempting to upgrade again:
invalid descriptor: db_name.public.table_name (87) because 'mutation job 863082137902612481: job not found'
SQLSTATE: 55000


[root@f4236dc6da72 cockroach]# cockroach debug doctor examine cluster --insecure
Examining 89 descriptors and 92 namespace entries...
  ParentID  52, ParentSchemaID 29: relation "table_name" (87): mutation job 863082137902612481: job 863082137902612481 not found
Examining 41937 jobs...
ERROR: validation failed
Failed running "debug doctor examine cluster"
[root@f4236dc6da72 cockroach]# 
```

The `show jobs;` had a lot of succeeded `MIGRATION`, `BACKUP`, and `SCHEMA CHANGE` jobs, as well as these two `reverting` jobs from before:

```
        job_id       |   job_type    |                                                                                                                    description                                                                                                                    | statement | user_name |  status   | running_status |          created           |          started           |          finished          |          modified          | fraction_completed |        error         | coordinator_id |      trace_id       |          last_run          |          next_run          | num_runs |                                                                                                                                                                                             execution_errors
---------------------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+-----------+-----------+----------------+----------------------------+----------------------------+----------------------------+----------------------------+--------------------+----------------------+----------------+---------------------+----------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  902973788233302017 | SCHEMA CHANGE | ALTER TABLE db_name.public.table_name ADD COLUMN column_name_2 UUID NOT NULL DEFAULT gen_random_uuid()                                                                                                                                            |           | root      | reverting | NULL           | 2023-09-25 10:02:11.117015 | 2023-09-25 10:02:11.815853 | NULL                       | 2023-10-15 09:28:29.315739 |                  0 | job canceled by user |              2 |  540219201967946643 | 2023-10-15 09:28:29.491527 | 2023-10-15 09:59:59.491527 |        6 | {reverting execution from '2023-10-15 09:05:09.151065' to '2023-10-15 09:05:09.19302' on 3 failed: schema change not first in line,reverting execution from '2023-10-15 09:12:49.485118' to '2023-10-15 09:12:49.656985' on 1 failed: schema change not first in line,reverting execution from '2023-10-15 09:28:29.491527' to '2023-10-15 09:28:29.566428' on 3 failed: schema change not first in line}
  869008109992148993 | SCHEMA CHANGE | ALTER TABLE db_name.public.table_name DROP COLUMN column_name_1                                                                                                                                                                                   |           | root      | reverting | NULL           | 2023-05-28 10:43:48.844502 | 2023-05-28 10:43:49.743624 | NULL                       | 2023-10-15 09:28:29.315738 |                  0 | job canceled by user |              2 | 4815928956969527583 | 2023-10-15 09:28:29.488381 | 2023-10-15 09:59:59.488381 |        6 | {reverting execution from '2023-10-15 09:05:09.148601' to '2023-10-15 09:05:09.195941' on 3 failed: schema change not first in line,reverting execution from '2023-10-15 09:12:49.485007' to '2023-10-15 09:12:49.658647' on 1 failed: schema change not first in line,reverting execution from '2023-10-15 09:28:29.488381' to '2023-10-15 09:28:29.55871' on 3 failed: schema change not first in line}
```

I didn't know how to fix the issue, so I took these steps:
1. Backup.
1. I tried copying the data as a new table: `CREATE TABLE table_name_copy AS TABLE table_name;`. Later I found out that this command does not copy the data or even the schema!
2. Drop constraints and foreign keys.
3. Drop table.
4. Finalizing the update with `SET CLUSTER SETTING version = "22.1"`.
5. Trying `CREATE TABLE table_name AS TABLE table_name_copy;` and getting `ERROR: relation "table_name_copy" does not exist`!!!
6. Recreating the table. Inserting the data from the backup.
7. `ALTER TABLE` to `ADD CONSTRAINT` and `VALIDATE CONSTRAINT`.

Continue the updates and Done.
